### PR TITLE
[dns-sd] Untangle dependencies between Resolver and ResolverProxy

### PR DESCRIPTION
**Pull request by Damian-Nordic**
The relationship between Resolver and Resolver Proxy is unclear as both classes depend on each other and Resolver Proxy is implemented in a source file of concrete Resolver implementation. This makes it impossible to build Matter library with several Resolver implementations and select one to be used at runtime.

The main reason for introducing Resolver Proxy was the need for running multiple parallel DNS-SD queries. Address this by adding Discovery Delegate parameter to the Resolver interface methods for starting and stopping the discovery.

Then implement Resolver Proxy as a convenience class on top of the Resolver interface. Additionally, simplify the Commissionable Node Controller class by injecting a test Resolver into Resolver Proxy.

This change is sort of aligned with https://github.com/project-chip/connectedhomeip/pull/29264 that removed the usage of `ResolverProxy` from the operational discovery path.

### CODE REVIEW

File: src/lib/shell/commands/Dns.cpp [hunk 1]
```diff
@@ -211,31 +210,29 @@ bool ParseSubType(int argc, char ** argv, Dnssd::DiscoveryFilter & filter)
...
 {
     Dnssd::DiscoveryFilter filter;
+    VerifyOrReturnError(ParseSubType(argc, argv, filter), CHIP_ERROR_INVALID_ARGUMENT);
 
-    if (!ParseSubType(argc, argv, filter))
-    {
-        streamer_printf(streamer_get(), "Invalid argument\r\n");
-        return CHIP_ERROR_INVALID_ARGUMENT;
-    }
+    Dnssd::ResolverProxy resolverProxy;
```
Comment (andy31415) [inline src/lib/shell/commands/Dns.cpp:215]: Could we add some usage comments on ResolverProxy lifetimes and usage?

My first thought on this was that we use a resolverProxy on the stack to initiate an async operation, so I thought that seems dangerous. 

To me it looks like:
   - Init creates a `mDelegate = Platform::New<DiscoveryDelegate>();`
   - SetCommissioningDelegate sets `mCommissioningDelegate = delegate;`
   - DiscoverCommissionableNodes uses `mDelegate`
   - Destructor:
       - does `mDelegate->SetCommissioningDelegate(nullptr)`  <- This seems dodgy, do we affect async operation?
       - does mDelegate->Release, so I assume this frees memory
       
What I am wondering: what makes using a stack resolver proxy safe to use for async operations, given that resolverproxy has an internal delegate and logic.

I always struggled to even understand ResolverProxy. Maybe now would be a time to explain its intent and usage in comments.

Comment (Damian-Nordic) [reply src/lib/shell/commands/Dns.cpp:215]: I guess a part of the confusion is a bit unfortunate naming of `ReferenceCounter` class methods as `Release` only means `refcount--`, not the actual release.

Comment (andy31415) [reply src/lib/shell/commands/Dns.cpp:215]: So if I inline the ResolverProxy here, does this look right?

```cpp
// Constructor
Resolver *mResolver = &Resolver::Instance();
// Init call
mResolver.Init(DeviceLayer::UDPEndpointManager()); // we global re-init for every call. Is this ok?
DiscoveryDelegate *mDelegate = Platform::New<DiscoveryDelegate>();
// setter
mDelegate->SetCommissioningDelegate(&sDnsShellResolverDelegate);
// DiscoverCommissionableNodes
mResolver->DiscoverComissionableNodes(filter, *mDelegate);

// Destructor/Shutdown
mDelegate->SetCommissioningDelegate(nullptr); // this changes mDelegate before resolver completed
mDelegate->Release()
```

Comment (andy31415) [reply src/lib/shell/commands/Dns.cpp:215]: My complains are: resolver on stack always re-initializes the global resolver, which may be a bit ugly (not sure if we care though).

If we clear commissioningDelegate before the async resolution fails, how does the resolver know where to send the results? How do they get to sDnsShellResolverDelegate in this case, this part is unclear to me.

Comment (Damian-Nordic) [reply src/lib/shell/commands/Dns.cpp:215]: You inlined it correctly, and good catch - the complains seem valid. Let me fix that!

Comment (Damian-Nordic) [reply src/lib/shell/commands/Dns.cpp]: Fixed, added comments and renamed `DiscoveryDelegate` to `DiscoveryContext`.

**Review by andy31415 — APPROVED** [review]

**Pull request merged**